gce/survival-training/README-tf-estimator.md

```diff
@@ -0,0 +1,404 @@
+
+1. `--learning-rate` - Hyperparameter defining the size of the mesh with respect to which we perform our optimization.
+
+1. `--batch-norm-decay` - Hyperparameter specifying the time-discount in moving average of batch means in batch normalization.
+
+1. `--batch-norm-epsilon` - Hyperparameter for batch normalization which represents a slight increase to the batch variance estimate.
+
+
+We can specify most of these parameters right now. We will run the following command for training:
+
+```bash
+python cifar10_main.py \
+    --data-dir <TBD> \
+    --job-dir <TBD> \
+    --num-gpus 4 \
+    --train-steps 99999999 \
+    --momentum 0.9 \
+    --weight-decay 0.0002 \
+    --learning-rate 0.1 \
+    --batch-norm-decay 0.997 \
+    --batch-norm-epsilon 0.00001
+```
+
+For the hyperparameters, these are the default values provided by the trainer CLI, but we include them here explicitly so that it is easy to adapt this guide to training with different hyperparameters.
+
+Now we have to settle the matter of `--data-dir` and `--job-dir`.
+
+
+### Data
+
+The CIFAR-10 tutorial provides very friendly instructions for the generation of the training, evaluation, and validation datasets (as tfrecords). What we will do is:
+
+1. Download the dataset and generate the `.tfrecords` files locally
+
+1. Push them to a GCS bucket available to our GCE instances
+
+1. At training time, we will pass the GCS path to the data to the `--data-dir` argument
+
+To download the dataset and generate the `.tfrecords`, choose a local path at which you would like to do so, say `/tmp/cifar-10-data`. Then simply navigate to the [CIFAR-10 tutorial](https://github.com/tensorflow/models/tree/master/tutorials/image/cifar10_estimator) folder (you should have `git clone`d [tensorflow/models](https://github.com/tensorflow/models) to your local machine) and run:
+
+```bash
+python generate_cifar10_tfrecords.py --data-dir=/tmp/cifar-10-data
+```
+
+
+Now make a bucket in which to store this data (note that bucket names have to be globally unique, so we will prefix this with our project name):
+
+```bash
+gsutil mb gs://$(gcloud config get-value project)-cifar-10-data
+```
+
+Now we can upload our data to the bucket:
+
+```bash
+gsutil -m cp -r /tmp/cifar-10-data/*  gs://$(gcloud config get-value project)-cifar-10-data/
+```
+
+You can test that the data has been transferred using
+
+```bash
+gsutil ls gs://$(gcloud config get-value project)-cifar-10-data/
+```
+
+
+### Checkpoints
+
+In the course of training, checkpoints will be stored at the path provided to the trainer using the `--job-dir` argument. We can provide a GCS path for this argument, as well, and we will do so.
+
+Let us make ourselves a checkpoint bucket:
+
+```bash
+gsutil mb gs://$(gcloud config get-value project)-cifar-10-checkpoints
+```
+
+With all this preparation in place, we are ready to specify a startup script and define our instance metadata.
+
+
+## Cloud Source Repositories
```

Comment:
We could have it living on GCS and on the persistent disk. But I actually prefer the source repo version for the added tagging semantics that it brings to the table. I wouldn't mind adding a second option which uses GCS, though, but as a follow up PR.

Reply:
What is the tagging semantics?

If that is important/essential to the presented solution then it should be highlighted; if not, then it could be a distraction from the core parts of the solution.

Reply:
Responded in the post where I talked about switching to GitHub and GCS - it's not crucial.